src/components/implementation/voter/voter_src/rust_src/src/voter/voter_lib.rs

```diff
@@ -0,0 +1,207 @@
+	}
+}
+
+impl Component {
+	pub fn new(num_replicas: usize, sl: Sl, thd_entry: fn(sl:Sl, rep_id: usize)) -> Component {
+		assert!(num_replicas <= MAX_REPS);
+		//create new Component
+		let mut comp = Component {
+			replicas: Vec::with_capacity(num_replicas),
+			num_replicas,
+			new_data:false,
+		};
+
+		//create replicas,start their threads,add them to the componenet
+		for i in 0..num_replicas {
+			let thd = sl.spawn(move |sl:Sl| {thd_entry(sl, i);});
+			comp.replicas.push(Replica::new(i,thd));
+		}
+
+		comp
+	}
+
+	pub fn wake_all(&mut self) {
+		//update state first to avoid replicas beginning new read/write
+		//while still in old read/write state
+		for replica in &mut self.replicas {
+			assert!(replica.state != ReplicaState::Init);
+			if replica.is_blocked() {replica.state_transition(ReplicaState::Processing);}
+		}
+
+		for replica in &mut self.replicas {
+			replica.thd.wakeup();
+		}
+	}
+
+	pub fn collect_vote(& mut self) -> VoteStatus {
+		println!("Collecting Votes");
+		let mut processing_rep_id = 0;
+		let mut num_processing = 0;
+		for replica in &self.replicas {
+			if replica.is_processing() {
+				num_processing += 1;
+				//rep id is only useful in the case where only 1 replica is still processing
+				//so it would only be set once in that case.
+				processing_rep_id = replica.rep_id;
+			}
+		}
+		//if any of the replicas are still processing bail.
+		if num_processing > 0 {return VoteStatus::Inconclusive(num_processing,processing_rep_id)}
+
+		//check the request each replica has made
+		if !self.validate_msgs() {
+			let faulted = self.find_faulted_msg();
+			assert!(faulted>-1);
+			return VoteStatus::Fail(faulted as usize);
+		}
+
+		return VoteStatus::Success;
+	}
+
+	pub fn validate_msgs(&self) -> bool {
+		//compare each message against the first to look for difference (handle detecting fault later)
+		let ref msg = &self.replicas[0].data_buffer;
+		for replica in &self.replicas {
+			if !compare_msgs(msg,&replica.data_buffer) {return false}
+		}
+
+		true
+	}
+
+	//todo return result
+	pub fn find_faulted_msg(&self) -> i16 {
+		//store the number of replicas that agree, and rep id of sender
+		let mut concensus: [u8; MAX_REPS] = [0; MAX_REPS];
+
+		//find which replica disagrees with the majority
+		for i in 0..self.num_replicas {
+			let msg_a = &self.replicas[i].data_buffer;
+			for j in 0..self.num_replicas {
+				if i == j {continue}
+
+				let msg_b = &self.replicas[j].data_buffer;
+
+				if compare_msgs(msg_a,msg_b) {
+					concensus[i] += 1;
+				}
+			}
+		}
+		//go through concensus to get the rep id that sent the msg with least agreement
+		let mut min:u8 = 4;
+		let mut faulted:i16 = -1;
+		for (rep,msg_votes) in concensus.iter().enumerate() {
+			if  *msg_votes < min {
+				min = *msg_votes;
+				faulted = rep as i16;
+			}
+		}
+		return faulted;
+	}
+}
+
+pub fn compare_msgs(msg_a:&[u8;voter_lib::BUFF_SIZE], msg_b:&[u8;voter_lib::BUFF_SIZE]) -> bool {
+	for i in 0..voter_lib::BUFF_SIZE {
```

Comment:
Again, might not want to iterate through the whole buffer.  Just the data that has been filled out.